Edits based on offline discussions

tschneidereit opened the pull request:
These changes are the result of an extended discussion between me and @nmatsakis.

The most important bits are
- the removal of array types in favor of StructType constructor overloads and `.array` constructors accompanying all struct types.
- changing from transparent- to opaque-by-default for struct types.
- documenting prototype chains.

tschneidereit committed:
Add links to TOC

tschneidereit committed:
Change struct types to be opaque by default

Fixes #2

tschneidereit committed:
Explain new scheme for array handling
tl;dr: Fixed-sized arrays are now just structs with indexed properties. Additionally, each struct type has an accompanying array constructor and prototype. Explicit ArrayTypes are gone.

Fixes #1

tschneidereit committed:
Changed all vars to consts or lets

`const` for type definitions, `let` for everything else.

tschneidereit committed:
Fill in section on prototypes

Ignores value types for now.

tschneidereit committed:
Primitive type casting clarifications

tschneidereit committed:
Clarify valid contents of structure argument to StructType

nikomatsakis commented on core.md:
````diff
@@ -89,10 +113,35 @@ struct would:
+
+```js
+const LineType = new StructType({0: PointType, 1: PointType});
+```
````
Maybe not exactly equivalent...What about the length property?

tschneidereit replied:
I changed the example to include a `length` field. I think it'd be a shame not to really make the two definitions equivalent, and really don't see too much of a reason not to.

nikomatsakis commented on core.md:
```diff
@@ -130,83 +179,79 @@ default can save a significant amount of memory, particularly if you
 
 The resulting object is called a *typed object*: it will have the
-fields specified in `LineType`. Each field will be initialized to an
+fields specified in `Line`. Each field will be initialized to an
```
S/Line/LineType? Or else change the example above?

nikomatsakis commented on core.md:
```diff
@@ -130,83 +179,79 @@ default can save a significant amount of memory, particularly if you
 
-var PointsType = PointType.array(2);
-var array = new PointsType([{x: 1, y: 2}, {x: 3, y: 4}]);
+let array = new PointType.array([{x: 1, y: 2}, {x: 3, y: 4}]);
```
Is this right? It seems inconsistent with what you wrote about array constructor above. I'm not sure this example belongs here anymore

nikomatsakis commented on core.md:
```diff
@@ -274,25 +319,25 @@ before, but with a different offset. The offset is now 16, because it
 which are 8 bytes each).
 
-Unlike arrays and structs, accessing a field of primitive type does
+Unlike for arrays and structs, accessing a field of primitive type does
```
Since we no longer have "array types" (that is, they are not type descriptors), this sentence could probably read "unlike for structs,..."

nikomatsakis commented on core.md:
```diff
@@ -371,23 +416,44 @@ placed into a weakmap.
 ## Prototypes
 
-FIXME
+All typed objects have an accompanying `prototype`. The `[[Prototype]]` of new instances of a type is set to that `prototype`. For struct types, the `prototype`'s own `[[Prototype]]` is immutably set to `StructType.prototype`. The `[[Prototype]]` of arrays of a struct type `Point`, instantiated using `new Point.array()`, is set to `Point.array.prototype`. That object's own `[[Prototype]]` is set  ...
```
S/typed objects/type descriptors/ have an accompanying prototype?

tschneidereit replied:
I changed this to `type definitions`, as that's used elsewhere and, I think, makes the most sense.

nikomatsakis commented on core.md:
````diff
@@ -371,23 +416,44 @@ placed into a weakmap.
+
+```js
+const Point({x: float64, y: float64});
+const Line({start: Point, end: Point});
````
= new StructType (...)?

nikomatsakis commented on core.md:
```diff
@@ -371,23 +416,44 @@ placed into a weakmap.
+points2.__proto__ === points1.__proto__;
+Point.prototype.__proto__ === StructType.prototype;
+Point.array.prototype.__proto__ === StructType.array.prototype;
+line.__proto__ === Line.prototype;
```
Also StructType.array.prototype.**proto** === array.prototype?

tschneidereit replied:
I added this, yes. The reason I didn't before was that I was on the fence about it. I've come down on the "yes" side of the fence now, though. The only question is whether there should be a common base for value and struct types and their arrays, but that's for a later change.

nikomatsakis commented on core.md:
```diff
@@ -371,23 +416,44 @@ placed into a weakmap.
+constructor to create instances of struct type definitions or their accompanying
+array types, which in turn implies that we create a new backing buffer. Sometimes,
 though, it can be useful to take an existing array buffer and create a
 typed view onto its contents. That can be achieved using the `view`
```
Clarify that this only works for transparent types...?

nikomatsakis commented on core.md:
```diff
@@ -397,40 +463,27 @@ functions (but see section on opacity below):
+let opaquePoint = Point.view(buffer(point), offset(point));
+
+buffer(opaquePoint); // yields undefined
+offset(opaquePoint); // yields undefined
```
Presumably all these names should be in some module?

nikomatsakis commented on core.md:
```diff
@@ -397,40 +463,27 @@ functions (but see section on opacity below):
+const TransparentPoint = new StructType({x: float64, y: float64}, {transparent: true});
+const Point = new StructType({x: float64, y: float64});
+let point = new TransparentPoint({x: 10, y: 100});
+let opaquePoint = Point.view(buffer(point), offset(point));
```
I don't think we want to allow view with an opaque type. Besides th e security concerns, it will impair optimization, since we no longer have aliasing guarantees.

nikomatsakis commented:
Left some comments. Mostly minor but for the concern about opaque objects and view method.

tschneidereit committed:
Address @nmatsakis' feedback

Importantly, changes the part on views to make clear only transparent types can be created as views onto preexisting buffers.

The rest is largely bugfixes.

tschneidereit committed:
Clarifications and small fixes to core explainer

tschneidereit commented:
@nikomatsakis thanks for the comments. You're entirely right wrt the transparency and view method thing - fixed.

I also made some more slight changes and started work on the value types explainer. More to come for that, but I'll open a separate PR.

tschneidereit committed:
Explain struct type arrays as typed arrays

They're closely related to the existing typed arrays both in intent and in behavior.

nikomatsakis commented:
Based on discussion with @tschneidereit, going to close this PR for now pending revisions.
